### PR TITLE
[receiver/statsd] Add an exponential histogram option 

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** 
Adds a "histogram" option to enable the OTLP v0.11 auto-scaling exponential histogram aggregator for Timing and Histogram measurements. 

This uses the Histogram mapping functions and data structure implemented in github.com/lightstep/go-expohisto.

The same mapping functions and data structure are pending as for OTel-Go (e.g., https://github.com/open-telemetry/opentelemetry-go/pull/3022); if OTel-Go agrees to expose these APIs, this module can be updated to use it directly, see:

https://github.com/open-telemetry/opentelemetry-go/issues/3169
https://github.com/open-telemetry/opentelemetry-go/issues/3170

**Link to tracking Issue:** 
Fixes #5742

**Testing:** New tests added. 

**Documentation:** README.md updated with examples including histogram configuration.

### CODE REVIEW

File: receiver/statsdreceiver/go.mod
Comment (mx-psi) [inline receiver/statsdreceiver/go.mod:3]: Blocking this until #12916 is fixed

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (mx-psi) [inline receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: I don't think we should use testbed here: testbed is an environment for end to end tests, not unit tests. If you need this functionality, we should move it to a shared internal module.

**Review by mx-psi — DISMISSED** [review]

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: Are there similar utilities elsewhere I should find and use, instead? It's quite difficult to write a good test without helpers of this nature to print what's actually not matching correctly. Alternatively, is there an explainer for how I can properly use this code to test the receiver here? This library looks perfect for my use-case.

Comment (mx-psi) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: @djaglowski has done some work on testbed so can probably help us here. IMO we should move this diff function to an internal shared module, which shouldn't be a lot of work

File: receiver/statsdreceiver/go.mod
Comment (mx-psi) [reply receiver/statsdreceiver/go.mod:3]: This is fixed

**Comment by mx-psi** [comment]
This probably needs a rebase

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: @djaglowski how would you like to see the testbed code factored? The PR description explains what I did:
> Extended the testbed framework for exponential histogram testing; fixed a TODO in that file -- metric attributes were not being verified and now are.

File: receiver/statsdreceiver/testdata/config.yaml
Comment (codeboten) [inline receiver/statsdreceiver/testdata/config.yaml:10]: Should this be updated to histogram?

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (codeboten) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: Could this be moved to `internal/coreinternal/emtricstestutil`? Ideally in a separate PR that we can get merged pretty quickly

File: receiver/statsdreceiver/config_test.go
Comment (jmacd) [reply receiver/statsdreceiver/config_test.go]: Fixed w/ new tests.

File: receiver/statsdreceiver/go.mod
Comment (jmacd) [reply receiver/statsdreceiver/go.mod]: This resulted from the testbed import, which explains why that was a bad idea, e.g.,

github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver/protocol
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/statsdreceiver/protocol.test
github.com/open-telemetry/opentelemetry-collector-contrib/testbed/correctnesstests/metrics
github.com/open-telemetry/opentelemetry-collector-contrib/testbed/correctnesstests
github.com/open-telemetry/opentelemetry-collector-contrib/testbed/datareceivers
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver
github.com/prometheus/prometheus/discovery/install
github.com/prometheus/prometheus/discovery/azure
github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-10-01/compute
github.com/Azure/go-autorest/autorest/date

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: I think so. This means some refactoring because the utilities here (which I found nice, and useful) are somehow tied with a Prometheus dependency, whereas all I wanted was the small utilities. I'll try.

File: receiver/statsdreceiver/testdata/config.yaml
Comment (jmacd) [reply receiver/statsdreceiver/testdata/config.yaml:10]: I had missed some testing for invalid configuration. This is improved to verify histogram configs are only present with histogram observer type, etc.

File: go.mod
Comment (jmacd) [reply go.mod:387]: Added "TEMPORARY" in the PR description and

```
	// Note: the following import from otel-launcher-go is being replaced by
	// an official copy of the same code in the otel-go repository.  This is
	// external import of a critical data structure is temporary until
	// https://github.com/open-telemetry/opentelemetry-go/pull/3022 merges.
```
to reassure readers.

**Comment by jmacd** [comment]
Will leave this in draft until the testbed refactoring is complete.

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go:28]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13428

**Comment by jmacd** [comment]
Having trouble with build dependencies; this may have to wait until OTel-Go 1.10 is released, which includes the recent boundary-inclusivity change. The `lightstep/sdk/metric/aggregator/histogram/structure` used in this PR also depends on that release and the same data structure will likely not be accepted/released until OTel-Go 1.11.

**Comment by jmacd** [comment]
Please take another look. This now depends on the latest release of Lightstep's metric SDK as well as an "orphaned" `sdk/metric/aggregator/exponential/mapping` package that was merged after OTel-Go v0.31.0 and will not be released. If this continues to cause trouble, I will consider making a downstream copy of the mapping functions to smooth this over (i.e.,., `otel-launcher-go/lightstep/sdk/metric/aggregator/histogram/mapping`).

File: receiver/statsdreceiver/README.md
Comment (mx-psi) [inline receiver/statsdreceiver/README.md:34]:
```suggestion
For `"summary`, the statsD receiver will aggregate to one OTLP summary metric for one metric description (the same metric name with the same tags). It will send percentile 0, 10, 50, 90, 95, 100 to the downstream.  The `"histogram"` setting selects an [auto-scaling exponential histogram configured with only a maximum size](https://github.com/lightstep/otel-launcher-go/tree/main/lightstep/sdk/metric/aggregator/histogram/structure), as shown in the example below.
```

File: receiver/statsdreceiver/go.mod
Comment (mx-psi) [inline receiver/statsdreceiver/go.mod:20]: AIU, this change is on the spec but not in any released version (https://github.com/open-telemetry/opentelemetry-specification/commit/ea017156740775ddf132fbd3085cb176101a2042 is only on main). 

On the last SIG, a proposal to block this kind of changes was raised (see #14122); it's fine (IMO, other maintainers opinion may differ) to use specification-conformant code from the SDK that has not yet been released, but it's not fine to if the specification itself has not been released

**Review by mx-psi — COMMENTED** [review]
I haven't had time to give a full review to this PR so I am only leaving a few odd comments I had pending so that you have some early feedback

File: receiver/statsdreceiver/go.mod
Comment (mx-psi) [reply receiver/statsdreceiver/go.mod:20]: Spec is now released

Comment (jmacd) [reply receiver/statsdreceiver/go.mod]: To avoid complications of this nature, I published a copy of the exponential histogram data structure and mapping functions at github.com/lightstep/go-expohisto. This PR is updated to use that repository's v1.0.0 release.

**Comment by jmacd** [comment]
This PR has been updated to use a new, separate copy of Lightstep's OTel exponential histogram mapping functions and data structure. Please take another look!

File: receiver/statsdreceiver/protocol/metric_translator.go
Comment (mx-psi) [inline receiver/statsdreceiver/protocol/metric_translator.go:143]: Maybe worth opening an issue about this on opentelemetry-collector if you have not done this already

File: receiver/statsdreceiver/protocol/statsd_parser.go
Comment (mx-psi) [inline receiver/statsdreceiver/protocol/statsd_parser.go:57]: We should not remove this: this is part of the public API and would be a breaking change to remove it. Maybe the `protocol` folder should be `internal` to the statsdreceiver module, but that should be a different PR.

File: receiver/statsdreceiver/config.go
Comment (mx-psi) [inline receiver/statsdreceiver/config.go:73]: Are we (reasonably) sure that the OTel Go implementation range of accepted sizes will cover this range too? Should we be more restrictive to avoid breaking changes in the future if we go with a more restricted range on OTel Go?

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (mx-psi) [inline receiver/statsdreceiver/protocol/statsd_parser_test.go:1025]: Is this comment still accurate? We are not using OTel Go but `github.com/lightstep/go-expohisto/mapping/logarithm`, so I am not sure if the spec comment applies or not

File: receiver/statsdreceiver/protocol/metric_translator.go
Comment (mx-psi) [inline receiver/statsdreceiver/protocol/metric_translator.go:113]: Does goexpohisto support min/max? (I assume it does not, but just double checking in case we can add it)

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (mx-psi) [inline receiver/statsdreceiver/protocol/statsd_parser_test.go:1053]: This also needs to be `PutStr` I think
```suggestion
		dp.Attributes().PutStr("mykey", "myvalue")
```

**Review by mx-psi — COMMENTED** [review]
Hopefully my last round of comments, generally looks great :)

File: receiver/statsdreceiver/config.go
Comment (jmacd) [reply receiver/statsdreceiver/config.go:72]: OTel-Go will not be accepting these, it's the consumers (e.g., other OTel collector processors & exporters and their downstream receivers). There is not a specified limit to this number and we could remove this check. This limit is 100x the SDK-specified default size--if the OTLP protocol were changed to specify a limit, I suppose we'd update this code.

File: receiver/statsdreceiver/factory_test.go
Comment (jmacd) [reply receiver/statsdreceiver/factory_test.go:79]: Added.

File: receiver/statsdreceiver/protocol/metric_translator.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/metric_translator.go:113]: yes, it does. this PR dates to a time before the OTLP protocol had added it in the collector (due to the Gogo block), oops! Fixed (with new test apparatus).

File: receiver/statsdreceiver/protocol/metric_translator.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/metric_translator.go:119]: Will do.

**Review by jmacd — COMMENTED** [review]
This hasn't been updated for the v0.62 changes yet, otherwise mostly addressed your feedback @mx-psi/

File: receiver/statsdreceiver/protocol/statsd_parser_test.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/statsd_parser_test.go]: This was fixed, now removed.

File: receiver/statsdreceiver/protocol/statsd_parser.go
Comment (jmacd) [reply receiver/statsdreceiver/protocol/statsd_parser.go:57]: Will do.

**Comment by jmacd** [comment]
Updated for v0.62.0.

File: receiver/statsdreceiver/config.go
Comment (mx-psi) [reply receiver/statsdreceiver/config.go:72]: Not sure I follow here, do you mean that the OpenTelemetry Go SDK will not impose a maximum/minimum size for ExponentialHistograms then?

**Review by mx-psi — APPROVED** [review]
LGTM, my only pending question is https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12951#discussion_r994750957 (also one small thing missing from the v0.62.0 bump).

I will merge after you reply @jmacd  if @codeboten has no further comments

File: receiver/statsdreceiver/config.go
Comment (jmacd) [reply receiver/statsdreceiver/config.go:72]: As for minimum size, there's no point in a histogram with less than 2 buckets, so I set that as the minimum max-size. 

There isn't an OTel-specified maximum in the protocol or the SDK specification. I'm not opposed to such a limit being specified, I'm just not sure it's very important to do. An analogous situation in the trace SDK would be that we allow a limit on attributes count to be set but we don't limit the maximum number of attributes before dropping afaik. As a vendor, we have a realistic limit set for span attribute count and will reject spans with too many attributes, but trying to get everyone to agree to an upper bound seems difficult. The integer-width-imposed limit of 2**31 is much too high.

For the expo-histo specifically, we didn't place a limit on "scale" either. It's possible to have meaningful data at extreme scales with the existing protocol and someone might find a use for that. The underlying data structure sets a limit on scale at 20 simply because it's difficult to test beyond 20 (b/c even `math/big` has limits). Presumably we could extend the protocol to support 128-bit precision (i.e., greater max scale) or more than 2**31 entries (i.e,. greater max size) in the future.

File: receiver/statsdreceiver/go.mod
Comment (mx-psi) [inline receiver/statsdreceiver/go.mod:8]: Hm, I think you will need a `replace` statement for this module added on this `go.mod`

File: receiver/statsdreceiver/config.go
Comment (mx-psi) [reply receiver/statsdreceiver/config.go:72]: Thanks for the additional details. I think it's safe to assume then that we can keep support for the same range when we switch to use the OTel Go SDK :+1:

File: receiver/statsdreceiver/go.mod
Comment (jmacd) [reply receiver/statsdreceiver/go.mod:8]: I see, so I probably haven't really tested the min and max support yet. I'll add a bit more testing of the DiffMetrics().

Comment (jmacd) [reply receiver/statsdreceiver/go.mod:8]: See [3f95b9b](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12951/commits/3f95b9b2a1ae7eebdd962aec70294ba69a87c486). I'm concerned that these metrics test utilities are quite weak at testing and they could clearly be improved, but I also don't want to delay this work for want of testing infrastructure again (we did that once https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13428). Would you accept a new issue to cover improvements in testing utilities? (I've added TODOs to point out that the `MetricsDiff()` call here does not properly diff non-exponential histogram point Min and Max fields and the HasSum property).

Comment (mx-psi) [reply receiver/statsdreceiver/go.mod:8]: Yes, that is acceptable to me :+1:

Comment (jmacd) [reply receiver/statsdreceiver/go.mod:8]: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14934

File: internal/coreinternal/metricstestutil/metric_diff.go
Comment (codeboten) [inline internal/coreinternal/metricstestutil/metric_diff.go:209]: Please open a separate issue to track this

**Review by codeboten — APPROVED** [review]

File: internal/coreinternal/metricstestutil/metric_diff.go
Comment (jmacd) [reply internal/coreinternal/metricstestutil/metric_diff.go:209]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12951/files/5611572aed810144b005fd00f1fb668f93b17e5b#discussion_r994916101

**Pull request merged**
